kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -0,0 +1,193 @@
+        this.schema = Optional.of(newSchema); // will be verified as part of the build() call
+        return this;
+    }
+
+    @Override
+    public TransactionBuilder withPartitionColumns(Engine engine, List<String> partitionColumns) {
+        if (!partitionColumns.isEmpty()) {
+            this.partitionColumns = Optional.of(partitionColumns);
+        }
+        return this;
+    }
+
+    @Override
+    public Transaction build(Engine engine) {
+        SnapshotImpl snapshot;
+        try {
+            snapshot = (SnapshotImpl) table.getLatestSnapshot(engine);
+        } catch (TableNotFoundException tblf) {
+            logger.info(
+                    "Table {} doesn't exist yet. Trying to create a new table.",
+                    table.getPath(engine));
+            schema.orElseThrow(() -> new IllegalArgumentException(
+                    "Table doesn't exist yet. Must provide a new schema."));
+            // Table doesn't exist yet. Create an initial snapshot with the new schema.
+            Metadata metadata = getInitialMetadata();
+            Protocol protocol = getInitialProtocol();
+            LogReplay logReplay = getEmptyLogReplay(engine, metadata, protocol);
+            snapshot = new InitialSnapshot(
+                    table.getDataPath(),
+                    logReplay,
+                    metadata,
+                    protocol);
+        }
+
+        boolean isNewTable = snapshot.getVersion(engine) < 0;
+        validate(engine, snapshot, isNewTable);
+
+        return new TransactionImpl(
+                isNewTable,
+                table.getDataPath(),
+                table.getLogPath(),
+                snapshot,
+                engineInfo,
+                operation,
+                snapshot.getProtocol(),
+                snapshot.getMetadata());
+    }
+
+    /**
+     * Validate the given parameters for the transaction.
+     */
+    private void validate(Engine engine, SnapshotImpl snapshot, boolean isNewTable) {
+        // Validate the table has no features that Kernel doesn't yet support writing into it.
+        TableFeatures.validateWriteSupportedTable(
+                snapshot.getProtocol(),
+                snapshot.getMetadata(),
+                snapshot.getMetadata().getSchema());
+
+        if (!isNewTable) {
+            if (schema.isPresent()) {
+                throw new TableAlreadyExistsException(
+                        table.getPath(engine),
+                        "Table already exists, but provided a new schema. "
+                                + "Schema can only be set on a new table.");
+            }
+            if (partitionColumns.isPresent()) {
+                throw new TableAlreadyExistsException(
+                        table.getPath(engine),
+                        "Table already exists, but provided new partition columns. "
+                                + "Partition columns can only be set on a new table.");
+            }
```

Comment:
Fine for now because we are revisiting these create table APIs. But I feel like these should be more user-facing depending on how we define our connector expectations. As in, if connectors only use withSchema or withPartitionColumns to create new tables (we define this as their behavior) then this should probably just be a generic `TableAlreadyExistsException` that can be more user facing. Referencing details like "provided new partition columns" seems like a connector/dev facing message.

But I think this will change, as in if we provide methods like `createTableTransactionBuilder` we can clearly just throw a `TableAlreadyExistsException` in that scenario.